pkg/pillar/cmd/zedagent/handlecertconfig.go

```diff
@@ -0,0 +1,152 @@
+	"io/ioutil"
+
+	//zcert "github.com/lf-edge/eve/api/go/certs"
+	zconfig "github.com/lf-edge/eve/api/go/config"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	log "github.com/sirupsen/logrus"
+)
+
+// handler for controller cert object triggers
+func handleControllerCertModify(ctxArg interface{}, key string,
+	configArg interface{}) {
+	log.Infof("handleControllerCertModify(%s)\n", key)
+	ctx := ctxArg.(*zedagentContext)
+	config := configArg.(types.ControllerCertificate)
+	handleControllerCertUpdate(ctx.getconfigCtx, config, false)
+	log.Debugf("handleControllerCertModify(%s) done %v\n", key, config)
+}
+
+func handleControllerCertDelete(ctxArg interface{}, key string,
+	configArg interface{}) {
+	log.Infof("handleControllerCertDelete(%s)\n", key)
+	ctx := ctxArg.(*zedagentContext)
+	config := configArg.(types.ControllerCertificate)
+	handleControllerCertUpdate(ctx.getconfigCtx, config, true)
+	log.Debugf("handleControllerCertDelete(%s) done\n", key)
+}
+
+// controller cert update, triggers cipher context update
+func handleControllerCertUpdate(ctx *getconfigContext,
+	controllerCert types.ControllerCertificate, reset bool) {
+	//switch controllerCert.Type {
+	//case zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE:
+	//	updateCipherContextsWithControllerCert(ctx, controllerCert, reset)
+	//	return
+	//case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
+	//case zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE:
+	//	// TBD:XXX add appropriate handlers
+	//}
+	updateCipherContextsWithControllerCert(ctx, controllerCert, reset)
+	return
+
+}
+
+// update the cipher context with the controller cert
+func updateCipherContextsWithControllerCert(ctx *getconfigContext,
+	controllerCert types.ControllerCertificate, reset bool) {
+	log.Infof("%v, update cipher contexts\n", controllerCert.Key())
+	items := ctx.pubCipherContextConfig.GetAll()
+	for _, item := range items {
+		cipherCtx := item.(types.CipherContext)
+		// hash algorithm do not match
+		//if cipherCtx.HashScheme.String() !=
+		//	controllerCert.HashAlgo.String() ||
+		//	!bytes.Equal(cipherCtx.ControllerCertHash,
+		//		controllerCert.CertHash) {
+		//	continue
+		//}
+		if !bytes.Equal(cipherCtx.ControllerCertHash,
+			controllerCert.CertHash) {
+			continue
+		}
+		log.Infof("%v, updating ciphercontext, %s\n",
+			controllerCert.Key(), cipherCtx.Key())
+		if reset {
+			if len(cipherCtx.ControllerCert) != 0 {
+				cipherCtx.ControllerCert = []byte{}
+				publishCipherContextConfig(ctx, cipherCtx)
+			}
+			continue
+		}
+		cipherCtx.ControllerCert = controllerCert.Cert
+		publishCipherContextConfig(ctx, cipherCtx)
+	}
+}
+
+func getControllerCert(ctx *zedagentContext, suppliedHash []byte) []byte {
+	log.Infof("%v, get controller cert\n", suppliedHash)
+	items := ctx.subControllerCertConfig.GetAll()
+	for _, item := range items {
+		certConfig := item.(types.ControllerCertificate)
+		if bytes.Equal(certConfig.CertHash, suppliedHash) {
+			return certConfig.Cert
+		}
+	}
+	return []byte{}
```

Comment:
Better to return a []byte, error, or alternatively []byte, bool, where the boolean indicates whether it was found or not.

Reply:
done.